src/nrniv/neuronapi.cpp

```diff
@@ -0,0 +1,519 @@
+};
+
+struct SymbolTableIterator {
+    SymbolTableIterator(Symlist*);
+    char const* next(void);
+    int done(void);
+
+  private:
+    Symbol* current;
+};
+
+/****************************************
+ * Connections to the rest of NEURON
+ ****************************************/
+extern int nrn_nobanner_;
+extern int diam_changed;
+extern int nrn_try_catch_nest_depth;
+extern "C" void nrnpy_set_pr_etal(int (*cbpr_stdoe)(int, char*), int (*cbpass)());
+int ivocmain_session(int, const char**, const char**, int start_session);
+void simpleconnectsection();
+extern Object* hoc_newobj1(Symbol*, int);
+extern void nrn_change_nseg(Section*, int);
+extern std::tuple<int, char**> nrn_mpi_setup(int argc, char** argv);
+
+extern "C" {
+
+/****************************************
+ * Initialization
+ ****************************************/
+
+int nrn_init(int argc, const char** argv) {
+    nrn_nobanner_ = 1;
+    auto [final_argc, final_argv] = nrn_mpi_setup(argc, const_cast<char**>(argv));
+    errno = 0;
+    return ivocmain_session(final_argc, (const char**) final_argv, nullptr, 0);
+}
+
+void nrn_stdout_redirect(int (*myprint)(int, char*)) {
+    // the first argument of myprint is an integer indicating the output stream
+    // if the int is 1, then stdout, else stderr
+    // the char* is the message to display
+    nrnpy_set_pr_etal(myprint, nullptr);
+}
+
+/****************************************
+ * Sections
+ ****************************************/
+
+Section* nrn_section_new(char const* const name) {
+    // TODO: check for memory leaks; should we free the symbol, pitm, etc?
+    Symbol* symbol = new Symbol;
+    symbol->name = strdup(name);
+    symbol->type = 1;
+    symbol->u.oboff = 0;
+    symbol->arayinfo = 0;
+    hoc_install_object_data_index(symbol);
+    hoc_Item* itm;
+    new_sections(nullptr, symbol, &itm, 1);
+    return itm->element.sec;
+}
+
+void nrn_section_connect(Section* child_sec, double child_x, Section* parent_sec, double parent_x) {
+    nrn_pushsec(child_sec);
+    hoc_pushx(child_x);
+    nrn_pushsec(parent_sec);
+    hoc_pushx(parent_x);
+    simpleconnectsection();
+}
+
+void nrn_section_length_set(Section* sec, const double length) {
+    // TODO: call can_change_morph(sec) to check pt3dconst_; how should we handle
+    // that?
+    // TODO: is there a named constant so we don't have to use the magic number 2?
+    sec->prop->dparam[2] = length;
+    // nrn_length_change updates 3D points if needed
+    nrn_length_change(sec, length);
+    diam_changed = 1;
+    sec->recalc_area_ = 1;
+}
+
+double nrn_section_length_get(Section const* sec) {
+    return section_length(const_cast<Section*>(sec));
+}
+
+double nrn_section_Ra_get(Section const* sec) {
+    return nrn_ra(const_cast<Section*>(sec));
+}
+
+void nrn_section_Ra_set(Section* sec, double const val) {
+    // TODO: ensure val > 0
+    // TODO: is there a named constant so we don't have to use the magic number 7?
+    sec->prop->dparam[7] = val;
```

Comment:
At least I would add comment about what's this magic number 7! An enum or something similar in common header would be better!

Reply:
@ramcdougal pls, I have no idea for this one